File: web/lib/db/schema.ts
import {
  boolean,
  index,
  integer,
  jsonb,
  pgTable,
  real,
  text,
  timestamp,
  uniqueIndex,
  uuid,
} from "drizzle-orm/pg-core";
import type { CanonicalEntity } from "@/types/entity";
import type { SurvivorshipRule } from "@/types/rules";
import type { RunStats, RunStatus } from "@/types/run";
import type {
  ConflictClass,
  MatchMethod,
  MatchSignals,
  Severity,
} from "@/types/conflict";
import type { AuditAction, ResolutionStatus } from "@/types/resolution";
import type { SourceConfig, SourceKind } from "@/types/source";
import type { AuthMethod } from "@/types/credentials";

/**
 * The single source of truth for Canon's schema. Drizzle owns it; the engine
 * reads and writes the same Postgres but never migrates it.
 *
 * Two writers, split by TABLE OWNERSHIP — this is what keeps runs
 * reproducible, and it is not enforced by the network topology:
 *
 *   sources             web only            (engine must never write)
 *   source_credentials  web only            (engine reads and decrypts; never writes)
 *   rulesets    web only                    (engine must never write)
 *   runs        web creates, engine updates status
 *   entities    engine only                 (web must never write)
 *   matches     engine only                 (web must never write)
 *   conflicts   engine only                 (web must never write)
 *   resolutions engine proposes, web records human review
 *   audit_log   both, append-only           (never updated, never deleted)
 */

/** A configured system Canon reads from. Non-secret connection shape only. */
export const sources = pgTable("sources", {
  id: uuid("id").primaryKey().defaultRandom(),
  name: text("name").notNull(),
  kind: text("kind").$type<SourceKind>().notNull(),
  /**
   * NEVER credentials. Object names, table names, filters — nothing else.
   * Secrets live one table down in `source_credentials`, encrypted, and there
   * is no field here a password could travel in even by accident.
   */
  config: jsonb("config").$type<SourceConfig>().notNull(),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
});

/**
 * One source's credentials. The only table in Canon that holds a secret.
 *
 * This replaced the per-integration keys that used to sit in the engine's
 * `.env` (`SALESFORCE_PASSWORD`, `DATABRICKS_TOKEN`, …). Those could not
 * express two Salesforce orgs, could not be rotated without a deploy, and gave
 * every run the same identity regardless of who started it.
 *
 * The AGENTS.md rule they existed to satisfy is intact and is in fact stronger
 * here: a credential is still never in `sources.config`, is still never
 * readable by the web layer's own UI, and now additionally is never readable by
 * anyone holding only a database dump. What moved is the storage location, not
 * the trust boundary.
 *
 *   ciphertext   AES-256-GCM over the JSON secret bundle
 *   iv           96-bit nonce, fresh per write, never reused under a key
 *   authTag      128-bit GCM tag — tamper-evidence, not just confidentiality
 *   keyVersion   which root key this row was sealed under, so rotation is
 *                a background re-encrypt rather than a flag day
 *
 * Written by the web layer only. The engine decrypts and reads; it must never
 * write here, the same way it must never write `sources` — a run whose
 * credentials changed underneath it is a run that cannot be reproduced.
 */
export const sourceCredentials = pgTable(
  "source_credentials",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    sourceId: uuid("source_id")
      .notNull()
      .references(() => sources.id, { onDelete: "cascade" }),
    /** Which shape of `@/types/credentials` this row holds. */
    method: text("method").$type<AuthMethod>().notNull(),
    /**
     * Non-secret connection shape: hostnames, HTTP paths, warehouse names. In
     * clear because the console displays them back and the OAuth authorize URL
     * is built from them.
     */
    publicValues: jsonb("public_values").$type<Record<string, string>>().notNull().default({}),
    /** Base64 AES-256-GCM ciphertext of the secret bundle. */
    ciphertext: text("ciphertext").notNull(),
    /** Base64 96-bit nonce. Fresh on every write — GCM nonce reuse is fatal. */
    iv: text("iv").notNull(),
    /** Base64 128-bit GCM auth tag. */
    authTag: text("auth_tag").notNull(),
    /** Root key generation. Lets a rotation re-encrypt row by row. */
    keyVersion: integer("key_version").notNull().default(1),
    /** OAuth only — when the access token dies and the refresh token is spent. */
    expiresAt: timestamp("expires_at", { withTimezone: true }),
    /** OAuth only — what the provider actually granted, which is not always what was asked. */
    scope: text("scope"),
    /**
     * The Clerk user whose grant this is. Answers "whose access did this run
     * use?", which is the credential half of the question `audit_log.actor`
     * answers for decisions.
     */
    connectedBy: text("connected_by").notNull(),
    /** Last successful connection, stamped by the engine's preflight. */
    lastVerifiedAt: timestamp("last_verified_at", { withTimezone: true }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    // One credential per source. Reconnecting replaces the row rather than
    // adding a second — two live grants for one source would make "which one
    // did the run use?" unanswerable.
    uniqueIndex("source_credentials_source_idx").on(table.sourceId),
  ],
);

/**
 * Survivorship rules, versioned. Never edited in place: a run records which
 * rulesetId it used, and mutable rules would make past runs unreproducible.
 */
export const rulesets = pgTable("rulesets", {
  id: uuid("id").primaryKey().defaultRandom(),
  name: text("name").notNull(),
  version: integer("version").notNull().default(1),
  rules: jsonb("rules").$type<SurvivorshipRule[]>().notNull(),
  isActive: boolean("is_active").notNull().default(false),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
});

/** One reconciliation execution. */
export const runs = pgTable("runs", {
  id: uuid("id").primaryKey().defaultRandom(),
  /** The CRM side. */
  sourceAId: uuid("source_a_id")
    .notNull()
    .references(() => sources.id),
  /** The warehouse side. */
  sourceBId: uuid("source_b_id")
    .notNull()
    .references(() => sources.id),
  rulesetId: uuid("ruleset_id")
    .notNull()
    .references(() => rulesets.id),
  status: text("status").$type<RunStatus>().notNull().default("queued"),
  stats: jsonb("stats").$type<RunStats>().notNull(),
  error: text("error"),
  startedAt: timestamp("started_at", { withTimezone: true }).notNull().defaultNow(),
  finishedAt: timestamp("finished_at", { withTimezone: true }),
});

/**
 * The parts of a CanonicalEntity that are not promoted to their own column.
 * externalId, entityType, parentExternalId and lastModifiedAt are columns
 * because matching and survivorship read them directly.
 */
export type EntityFields = Pick<
  CanonicalEntity,
  "name" | "normalizedName" | "attributes"
>;

/** Normalized records from both sides, per run. Written by the engine only. */
export const entities = pgTable(
  "entities",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    runId: uuid("run_id")
      .notNull()
      .references(() => runs.id),
    sourceId: uuid("source_id")
      .notNull()
      .references(() => sources.id),
    /** The id in the source system. */
    externalId: text("external_id").notNull(),
    entityType: text("entity_type").$type<CanonicalEntity["entityType"]>().notNull(),
    fields: jsonb("fields").$type<EntityFields>().notNull(),
    /** The hierarchy edge. */
    parentExternalId: text("parent_external_id"),
    /** Drives recency survivorship. */
    lastModifiedAt: timestamp("last_modified_at", { withTimezone: true }),
  },
  (table) => [index("entities_run_source_external_idx").on(table.runId, table.sourceId, table.externalId)],
);

/** A pair of entities believed to be the same real-world thing. */
export const matches = pgTable("matches", {
  id: uuid("id").primaryKey().defaultRandom(),
  runId: uuid("run_id")
    .notNull()
    .references(() => runs.id),
  entityAId: uuid("entity_a_id")
    .notNull()
    .references(() => entities.id),
  entityBId: uuid("entity_b_id")
    .notNull()
    .references(() => entities.id),
  /** 0..1 */
  confidence: real("confidence").notNull(),
  method: text("method").$type<MatchMethod>().notNull(),
  signals: jsonb("signals").$type<MatchSignals>().notNull(),
});

/** A field-level disagreement within a match. */
export const conflicts = pgTable("conflicts", {
  id: uuid("id").primaryKey().defaultRandom(),
  runId: uuid("run_id")
    .notNull()
    .references(() => runs.id),
  matchId: uuid("match_id")
    .notNull()
    .references(() => matches.id),
  field: text("field").notNull(),
  valueA: text("value_a").notNull(),
  valueB: text("value_b").notNull(),
  conflictClass: text("conflict_class").$type<ConflictClass>().notNull(),
  /** 1..4 — see AGENTS.md § Severity Model. Read by the auto-apply gate. */
  severity: integer("severity").$type<Severity>().notNull(),
  detectedBy: text("detected_by").$type<"rule" | "agent">().notNull(),
});

/** The proposed and final answer for a conflict. */
export const resolutions = pgTable("resolutions", {
  id: uuid("id").primaryKey().defaultRandom(),
  conflictId: uuid("conflict_id")
    .notNull()
    .references(() => conflicts.id),
  /** `validate` guarantees this is an observed value or a legal normalization. */
  proposedValue: text("proposed_value").notNull(),
  /** Plain-English, agent-authored. */
  rationale: text("rationale").notNull(),
  /** Which survivorship rule fired, if any. */
  appliedRuleId: text("applied_rule_id"),
  /** 0..1 */
  confidence: real("confidence").notNull(),
  status: text("status").$type<ResolutionStatus>().notNull().default("proposed"),
  reviewedBy: text("reviewed_by"),
  /** What the human chose instead. */
  overrideValue: text("override_value"),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  reviewedAt: timestamp("reviewed_at", { withTimezone: true }),
});

/** Append-only. Never updated, never deleted — the audit trail is the product. */
export const auditLog = pgTable("audit_log", {
  id: uuid("id").primaryKey().defaultRandom(),
  runId: uuid("run_id")
    .notNull()
    .references(() => runs.id),
  resolutionId: uuid("resolution_id").references(() => resolutions.id),
  action: text("action").$type<AuditAction>().notNull(),
  /** "engine" | "agent" | a reviewer identifier. */
  actor: text("actor").notNull(),
  detail: jsonb("detail").$type<Record<string, unknown>>().notNull(),
  createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
});

File: engine/pipeline/schema.py
"""The tool contract.

Mirrors ``web/types/`` exactly. Every payload crossing the layer boundary is a
model defined here — no bare dicts. AGENTS.md: *do not change one side without
changing the other in the same commit.*

The wire format is camelCase because the web layer speaks it; Python stays
snake_case. ``CanonModel`` carries that translation so no call site has to.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator
from pydantic.alias_generators import to_camel

# ── Enumerations (mirror the Zod enums in web/types/) ────────────────────────

EntityType = Literal["account", "user", "role", "territory"]
SourceKind = Literal["salesforce", "hubspot", "databricks", "snowflake", "synthetic"]
RuleStrategy = Literal["prefer_source", "most_recent", "most_complete", "escalate"]
#: Which model family the `propose` node calls. Recorded on the run row for the
#: same reason `ruleset_id` is: two runs that disagree are only comparable if
#: what produced each answer is on the record. Mirrored in
#: web/types/run.ts::modelProviderSchema.
ModelProvider = Literal["claude", "gemini"]
RunStatus = Literal[
    "queued", "extracting", "matching", "detecting", "resolving", "complete", "failed"
]
ConflictClass = Literal["structural", "attribute", "cosmetic", "orphan", "duplicate"]
#: ``orphan`` is not a matching technique — it is the absence of one. An entity
#: with no counterpart still needs a row in ``matches`` because ``conflicts``
#: requires a match to point at and the console's queue inner-joins through it;
#: that row is self-referential (both sides the same entity) with confidence 0.
#: Naming the method honestly is what keeps a reader from mistaking a
#: self-referential row for a real pairing.
#: Mirrored in web/types/conflict.ts::matchMethodSchema.
MatchMethod = Literal["exact_key", "embedding", "fuzzy", "agent", "orphan"]
ResolutionStatus = Literal[
    "proposed", "auto_applied", "escalated", "approved", "rejected", "overridden"
]
#: The engine writes the first five and the last; the web layer writes the
#: human_* three. ``human_rejected`` is not in AGENTS.md § Database Schema's
#: list — it has to exist, because a rejection recorded as any other action
#: would put a false statement in the one table that is beyond dispute.
#: Mirrored in web/types/resolution.ts::auditActionSchema.
#: ``run_degraded`` is the second action not in AGENTS.md § Database Schema's
#: original list, and it exists for the same reason ``human_rejected`` does: the
#: system has a state it could not otherwise name truthfully. A run whose model
#: calls failed COMPLETED — it is not ``run_failed`` — but its escalation count
#: is not the agent's judgement. Recording that as anything else, or not at all,
#: leaves the audit trail asserting a clean run.
AuditAction = Literal[
    "run_started",
    "conflict_detected",
    "resolution_proposed",
    "auto_applied",
    "escalated",
    "human_approved",
    "human_rejected",
    "human_overridden",
    "run_degraded",
    "run_failed",
]

#: 4 structural · 3 attribute (ownership) · 2 attribute (non-rollup) · 1 cosmetic.
#: An integer the auto-apply gate reads, not a hue — see AGENTS.md § Severity Model.
Severity = Literal[1, 2, 3, 4]

DetectedBy = Literal["rule", "agent"]


class CanonModel(BaseModel):
    """Base for every model on the boundary: camelCase on the wire, snake_case here."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="forbid",
    )


# ── Entities ─────────────────────────────────────────────────────────────────


class CanonicalEntity(CanonModel):
    """A record from either side, normalized. Mirrors web/types/entity.ts."""

    external_id: str
    entity_type: EntityType
    name: str
    #: Folded; what matching actually compares. Folding is recorded, not destructive.
    normalized_name: str
    #: The hierarchy edge.
    parent_external_id: str | None = None
    attributes: dict[str, str | None] = Field(default_factory=dict)
    #: ISO 8601 — drives recency survivorship.
    last_modified_at: str | None = None
    source_id: str


# ── Survivorship ─────────────────────────────────────────────────────────────


class SurvivorshipRule(CanonModel):
    """Mirrors web/types/rules.ts.

    Rules are evaluated most specific first: exact ``field`` + exact
    ``entity_type`` beats a wildcard on either. First match wins. If no rule
    matches, the conflict escalates — Canon never guesses in the absence of
    policy.
    """

    id: str
    #: ``"*"`` applies to any field.
    field: str
    entity_type: str
    strategy: RuleStrategy
    preferred_source_id: str | None = None
    #: Plain English — goes into the agent prompt.
    description: str

    @model_validator(mode="after")
    def _prefer_source_needs_a_source(self) -> SurvivorshipRule:
        if self.strategy == "prefer_source" and not self.preferred_source_id:
            raise ValueError("preferred_source_id is required when strategy is prefer_source")
        return self


def rule_specificity(rule: SurvivorshipRule) -> int:
    """Higher is more specific. Ties keep author order, so first match wins."""
    return (0 if rule.field == "*" else 2) + (0 if rule.entity_type == "*" else 1)


# ── Matching and detection ───────────────────────────────────────────────────


class MatchSignals(CanonModel):
    """What the combined confidence was computed from."""

    exact_key: bool
    cosine: float
    fuzzy_ratio: float


class Match(CanonModel):
    """A pair of entities believed to be the same real-world thing."""

    id: UUID | None = None
    run_id: UUID
    entity_a_id: UUID
    entity_b_id: UUID
    confidence: float = Field(ge=0.0, le=1.0)
    method: MatchMethod
    signals: MatchSignals


class Conflict(CanonModel):
    """A field-level disagreement within a match. Mirrors web/types/conflict.ts."""

    id: UUID | None = None
    run_id: UUID
    match_id: UUID
    field: str
    value_a: str
    value_b: str
    conflict_class: ConflictClass
    severity: Severity
    detected_by: DetectedBy


class Resolution(CanonModel):
    """The proposed and final answer for a conflict."""

    id: UUID | None = None
    conflict_id: UUID
    #: `validate` guarantees this is an observed value or a legal normalization.
    proposed_value: str
    #: Plain-English, agent-authored. A reviewer decides from this in five seconds.
    rationale: str
    #: Which survivorship rule fired, if any.
    applied_rule_id: str | None = None
    confidence: float = Field(ge=0.0, le=1.0)
    status: ResolutionStatus = "proposed"
    reviewed_by: str | None = None
    override_value: str | None = None
    created_at: datetime | None = None
    reviewed_at: datetime | None = None


# ── Run control (the HTTP surface in main.py) ────────────────────────────────


class RunStats(CanonModel):
    """Mirrors web/types/run.ts. Every field is a headline number in the console."""

    entities_a: int = 0
    entities_b: int = 0
    candidate_pairs: int = 0
    matches: int = 0
    conflicts: int = 0
    auto_resolved: int = 0
    escalated: int = 0
    tokens_used: int = 0

    # ── Agent health ─────────────────────────────────────────────────────────
    # A failed model call escalates. That is the correct behaviour, but it means
    # `escalated` silently mixes two different facts: conflicts the agent judged
    # genuinely ambiguous, and conflicts nobody ever got an answer for. Without
    # these three, a run throttled into escalating most of its work is
    # indistinguishable from a run that found that much real ambiguity — the
    # eval harness has always said so loudly, and the console could not say it
    # at all.
    #: Model requests that returned an answer.
    model_calls: int = 0
    #: Requests that failed and escalated instead. Every one of these is an
    #: escalation that is NOT a statement about the conflict.
    model_errors: int = 0
    #: Requests the provider throttled and that were retried. Cost wall-clock,
    #: not correctness — a run can be slow and still be entirely trustworthy.
    model_rate_limited: int = 0


class StartRunRequest(CanonModel):
    """POST {ENGINE_URL}/runs.

    Only the id crosses the wire. The engine reads sources and ruleset from
    Postgres using ``run_id``: passing configuration would create two sources of
    truth for what a run actually executed, which breaks reproducibility.
    """

    run_id: UUID


class StartRunResponse(CanonModel):
    run_id: UUID
    accepted: Literal[True] = True


class RunStatusResponse(CanonModel):
    """GET {ENGINE_URL}/runs/:runId/status — polled by the console."""

    run_id: UUID
    status: RunStatus
    stats: RunStats
    error: str | None = None


class ProviderStatus(CanonModel):
    """One model provider, and whether this engine can actually call it.

    ``configured`` reports the presence of a key, never the key. The console
    reads this to default the run form to a provider that will work and to say
    why the other one will not — the same reasoning that puts "not connected"
    next to a source with no stored credentials.
    """

    provider: ModelProvider
    #: The resolution model this engine would use. Named so the console shows
    #: what a run will execute rather than a family label.
    model: str
    configured: bool


class ProvidersResponse(CanonModel):
    """GET {ENGINE_URL}/providers — read by the new-run screen."""

    providers: list[ProviderStatus]
